### PR TITLE
Highlander Work So Far

### PR DESCRIPTION
**Pull request by fieldju**
PR for what has been done so far
Things to remember

1. Java Docs and updated unit tests and static code analysis coming next sprint.
2. This PR is to a branch and not master.
3. We can make sticky notes for TODO's

### CODE REVIEW

File: src/main/java/com/nike/cerberus/SecureDataAction.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import static com.nike.cerberus.record.RoleRecord.ROLE_READ;
+import static com.nike.cerberus.record.RoleRecord.ROLE_WRITE;
+import static io.netty.handler.codec.http.HttpMethod.GET;
+import static io.netty.handler.codec.http.HttpMethod.POST;
+
+public enum SecureDataAction {
+
+    READ(GET, ImmutableSet.of(ROLE_OWNER, ROLE_WRITE, ROLE_READ)),
+    WRITE(POST, ImmutableSet.of(ROLE_OWNER, ROLE_WRITE)),
+    DELETE(HttpMethod.DELETE, ImmutableSet.of(ROLE_OWNER, ROLE_WRITE));
```
Comment (sdford) [inline src/main/java/com/nike/cerberus/SecureDataAction.java]: nit: should import static here to be consistent

Comment (fieldju) [reply src/main/java/com/nike/cerberus/SecureDataAction.java]: You cannot because DELETE is already defined in the enum

File: src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ public AuthenticateIamPrincipal(final AuthenticationService authenticationServic
     private ResponseInfo<IamRoleAuthResponse> authenticate(RequestInfo<IamPrincipalCredentials> request) {
         final IamPrincipalCredentials credentials = request.getContent();
 
-        log.info("{}: {}, IAM Auth Event: the IAM principal {} with ip: {} in attempting to authenticate in region {}",
+        log.debug("{}: {}, IAM Auth Event: the IAM principal {} with ip: {} in attempting to authenticate in region {}",
```
Comment (sdford) [inline src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java]: I think this information is useful in production (as INFO level). It helps us troubleshoot customer issues.

File: src/main/java/com/nike/cerberus/service/AuthTokenService.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+        this.uuidSupplier = uuidSupplier;
+        this.tokenHasher = tokenHasher;
+        this.authTokenGenerator = authTokenGenerator;
+        this.authTokenDao = authTokenDao;
+        this.dateTimeSupplier = dateTimeSupplier;
+    }
+
+    @Transactional
+    public CerberusAuthToken generateToken(String principal,
```
Comment (sdford) [inline src/main/java/com/nike/cerberus/service/AuthTokenService.java]: Do we want Hystrix metrics around these methods? It seems like before we were wrapping the VaultClient with Hystrix

File: src/main/java/com/nike/cerberus/service/AuthenticationService.java [hunk 4]
```diff
@@ -268,19 +273,50 @@ private IamRoleAuthResponse authenticate(IamPrincipalCredentials credentials, Ma
         return iamRoleAuthResponse;
     }
 
+    private AuthTokenResponse createToken(String principal,
```
Comment (sdford) [inline src/main/java/com/nike/cerberus/service/AuthenticationService.java]: repeat about Hystrix metrics

File: src/main/java/com/nike/cerberus/SecureDataAction.java (hunk 1 above)
Comment (sdford) [reply src/main/java/com/nike/cerberus/SecureDataAction.java]: Got it

File: src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java (hunk 2 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java]: yeah, I need to revert that, it was quite noisy during the perf test.

File: src/main/java/com/nike/cerberus/service/AuthTokenService.java (hunk 3 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/service/AuthTokenService.java]: Todd has a ticket in the backlog or on his desk for Hystrix and RDS. Right now we don't have that integration

File: src/main/java/com/nike/cerberus/service/AuthenticationService.java (hunk 4 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/service/AuthenticationService.java]: Todd has a ticket in the backlog or on his desk for Hystrix and RDS. Right now we don't have that integration

**Comment by sdford** [comment]
It might be worth rebasing this branch with `master` before merging.

The first nine commits in this PR were squashed and merged into master already. The base highlander branch has the squashed commit as well.

File: src/main/java/com/nike/cerberus/domain/CerberusAuthToken.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+ */
+
+package com.nike.cerberus.domain;
+
+
+import com.nike.cerberus.PrincipalType;
+
+import java.time.OffsetDateTime;
+
+public class CerberusAuthToken {
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/domain/CerberusAuthToken.java]: One line of Javadoc is nice.

File: src/main/java/com/nike/cerberus/domain/SecureDataResponse.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    public boolean isRenewable() {
+        return renewable;
+    }
+
+    public void setRenewable(boolean renewable) {
+        this.renewable = renewable;
+    }
+
+    public int getLeaseDuration() {
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/domain/SecureDataResponse.java]: Javadoc should say what unit of time everywhere, there were bugs in the past in the clients because people got confused.

File: src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java (hunk 2 above)
Comment (tlisonbee) [reply src/main/java/com/nike/cerberus/endpoints/authentication/AuthenticateIamPrincipal.java]: yes, revert

File: src/main/java/com/nike/cerberus/endpoints/secret/DeleteSecureData.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.SecurityContext;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+public class DeleteSecureData extends SecureDataEndpointV1<Void, Object> {
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/DeleteSecureData.java]: Do we have principal and IP logging on this?  We need that everywhere on create/delete/update at minimum.

File: src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    private ResponseInfo<Object> listKeys(SecureDataRequestInfo info) {
+        Set<String> keys = secureDataService.listKeys(info.getPath());
+
+        if (keys.isEmpty()) {
+            return generateVaultStyleResponse(new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder().build(),
+                    HttpResponseStatus.NOT_FOUND.code());
+        }
+
+        SecureDataResponse response = new SecureDataResponse();
+        response.setRequestId(UUID.randomUUID().toString()); // maybe use trace id?
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java]: This seems weird.  Yes, maybe trace id so we can correlate it to something?

File: src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+        if (! data.isPresent()) {
+            return generateVaultStyleResponse(new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder().build(),
+                    HttpResponseStatus.NOT_FOUND.code());
+        }
+
+        SecureDataResponse response = new SecureDataResponse();
+        response.setRequestId(UUID.randomUUID().toString());
+        try {
+            response.setData(new ObjectMapper().readTree(data.get()));
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java]: Are we converting to JSON here?  Wasn't there one SDB where they had used Vault APIs directly and didn't do the normal key/value pairs?  Maybe we need the internals to support arbitrary files from day one?

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        if (! sdbId.isPresent() || ! permissionService.doesPrincipalHavePermission(principal, sdbId.get(),
+                        SecureDataAction.fromMethod(request.getMethod()))) {
+            log.info("SDB Id not found or permission was not granted for principal, principal: {}, sdb: {}," +
+                    " path: {}", principal.getName(), sdbId.orElseGet(() -> "missing"), request.getPath());
+            return generateVaultStyleResponse(longRunningTaskExecutor,
+                    ctx,
+                    new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder()
+                            .withError("permission denied")
+                            .build(),
+                    HttpResponseStatus.FORBIDDEN.code()
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: Are we changing error codes?  I think vault gave 404 to avoid information disclosure if someone doesn't have permissions.

I like 404 for non-authed users especially

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+        CerberusPrincipal principal = (CerberusPrincipal) securityContext.get().getUserPrincipal();
+
+        String sdbBasePath = String.format("%s/%s/", requestInfo.getCategory(), requestInfo.getSdbSlug());
+        Optional<String> sdbId = safeDepositBoxService.getSafeDepositBoxIdByPath(sdbBasePath);
+
+        // check that the principal has the proper permissions or throw Vault style permission denied error
+        if (! sdbId.isPresent() || ! permissionService.doesPrincipalHavePermission(principal, sdbId.get(),
+                        SecureDataAction.fromMethod(request.getMethod()))) {
+            log.info("SDB Id not found or permission was not granted for principal, principal: {}, sdb: {}," +
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: I think we should log specifically "if it wasn't found" OR "if there wasn't permissions"

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+        return doExecute(requestInfo, request, longRunningTaskExecutor, ctx, securityContext.get());
+    }
+
+    private boolean doesRequestHaveRequiredParams(String category, String sdbSlug) {
+        return isNotBlank(category) || isNotBlank(sdbSlug);
+    }
+
+    protected SecureDataRequestInfo parseInfoFromPath(String path) {
+        String[] parts = path.split("/", 6);
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: Probably should have named constants for all of these values.  And a unit test just to make sure nobody ever breaks.

File: src/main/java/com/nike/cerberus/endpoints/secret/WriteSecureData.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.SecurityContext;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+public class WriteSecureData extends SecureDataEndpointV1<Object, Object> {
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/secret/WriteSecureData.java]: I might be missing it but are we checking permissions with this action?  Also, we should have some logging with principal name and ip and path

File: src/main/java/com/nike/cerberus/jobs/LockingJob.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        log.info("Lock acquired for Job: {}, executing lockable code", jobName);
+        try {
+            executeLockableCode();
+        } catch (Throwable t) {
+            log.error("Failed to execute lockable job, releasing lock", t);
+        }
+
+        log.info("Attempting to release lock for Job: {}", jobName);
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/jobs/LockingJob.java]: Seems like this should go in a finally block

File: src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java (hunk 9 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java]: Yes this is taking the object structure and loading into a JsonNode, because the Object can be any valid json you can not cast it to a Map<String, String> it must be treated as generic JSON aka a JsonNode.

I manually tested it, and it works.

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java (hunk 10 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: You get a 403 if you try to access a path you do not have access to.
You get a 404 if you try to access a path you have access to but does not exist.

I manually tested both of those use cases against the existing API. Any deviations from this would be a breaking API change.

Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: Also if you have a bad token you also get a 403

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java (hunk 11 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: - [ ] separate log statement

File: src/main/java/com/nike/cerberus/jobs/LockingJob.java (hunk 14 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/jobs/LockingJob.java]: I was going to do that, but we want to be able to sleep between attempts.

File: src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java (hunk 8 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java]: - [ ] Use trace id

File: src/main/java/com/nike/cerberus/jobs/LockingJob.java (hunk 14 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/jobs/LockingJob.java]: - [x] move to finally block

File: src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java (hunk 10 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/endpoints/secret/SecureDataEndpointV1.java]: Adding link to this for the regression test ticket to verify these status codes for the edge cases.
